src/node_locks.cc

```diff
@@ -0,0 +1,738 @@
+  std::u16string resource_name(
+      reinterpret_cast<const char16_t*>(*resource_name_utf16),
+      resource_name_utf16.length());
+  String::Utf8Value client_id_utf8(isolate, args[1]);
+  std::string client_id(*client_id_utf8);
+  String::Utf8Value mode_utf8(isolate, args[2]);
+  std::string mode_str(*mode_utf8);
+  bool steal = args[3]->BooleanValue(isolate);
+  bool if_available = args[4]->BooleanValue(isolate);
+  Local<Function> callback = args[5].As<Function>();
+
+  Lock::Mode lock_mode =
+      mode_str == kSharedMode ? Lock::Mode::Shared : Lock::Mode::Exclusive;
+
+  Local<Promise::Resolver> waiting_promise;
+  if (!Promise::Resolver::New(context).ToLocal(&waiting_promise)) {
+    return;
+  }
+  Local<Promise::Resolver> released_promise;
+  if (!Promise::Resolver::New(context).ToLocal(&released_promise)) {
+    return;
+  }
+
+  args.GetReturnValue().Set(released_promise->GetPromise());
+
+  LockManager* manager = GetCurrent();
+  {
+    Mutex::ScopedLock scoped_lock(manager->mutex_);
+
+    // Register cleanup hook for the environment only once
+    if (manager->registered_envs_.insert(env).second) {
+      env->AddCleanupHook(LockManager::OnEnvironmentCleanup, env);
+    }
+
+    auto lock_request = std::make_unique<LockRequest>(env,
+                                                      waiting_promise,
+                                                      released_promise,
+                                                      callback,
+                                                      resource_name,
+                                                      lock_mode,
+                                                      client_id,
+                                                      steal,
+                                                      if_available);
+    // Steal requests get priority by going to front of queue
+    if (steal) {
+      manager->pending_queue_.emplace_front(std::move(lock_request));
+    } else {
+      manager->pending_queue_.push_back(std::move(lock_request));
+    }
+  }
+
+  manager->ProcessQueue(env);
+}
+
+void LockManager::Query(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+  HandleScope scope(isolate);
+  Local<Context> context = env->context();
+
+  Local<Promise::Resolver> resolver;
+  if (!Promise::Resolver::New(context).ToLocal(&resolver)) {
+    return;
+  }
+  args.GetReturnValue().Set(resolver->GetPromise());
+
+  Local<Object> result = Object::New(isolate);
+  Local<Array> held_list = Array::New(isolate);
+  Local<Array> pending_list = Array::New(isolate);
+  LockManager* manager = GetCurrent();
+
+  {
+    Mutex::ScopedLock scoped_lock(manager->mutex_);
+
+    uint32_t index = 0;
+    for (const auto& resource_entry : manager->held_locks_) {
+      for (const auto& held_lock : resource_entry.second) {
+        if (held_lock->env() == env) {
+          Local<Object> lock_info =
```

Comment:
You might consider moving the declaration for the `Local<Object> lock_info` to outside of the for loop and just reset it here so that we're reusing the same declaration on each iteration rather than creating a new one.